apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransaction.cs

```diff
@@ -0,0 +1,135 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Cloud.Spanner.V1;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Spanner.Data
+{
+    internal sealed class RetriableTransaction
+    {
+        private readonly SpannerConnection _connection;
+        private readonly IClock _clock;
+        private readonly IScheduler _scheduler;
+        private readonly RetriableTransactionOptions _options;
+
+        internal RetriableTransaction(SpannerConnection connection, IClock clock, IScheduler scheduler, RetriableTransactionOptions options = null)
+        {
+            _connection = GaxPreconditions.CheckNotNull(connection, nameof(connection));
+            _clock = GaxPreconditions.CheckNotNull(clock, nameof(clock));
+            _scheduler = GaxPreconditions.CheckNotNull(scheduler, nameof(scheduler));
+            _options = options ?? RetriableTransactionOptions.CreateDefault();
+        }
+
+        public async Task<TResult> RunAsync<TResult>(Func<SpannerTransaction, Task<TResult>> asyncWork, CancellationToken cancellationToken = default)
+        {
+            GaxPreconditions.CheckNotNull(asyncWork, nameof(asyncWork));
+
+            // Session will be initialized and subsequently modified by CommitAttempt.
+            PooledSession session = null;
+            try
+            {
+                return await ExecuteWithRetryAsync(CommitAttempt, cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                session?.Dispose();
+            }
+
+            async Task<TResult> CommitAttempt()
+            {
+                return await ExecuteHelper.WithErrorTranslationAndProfiling(
+                    async () =>
+                    {
+                        SpannerTransaction transaction = null;
+
+                        try
+                        {
+                            session = await (session?.WithFreshTransactionOrNewAsync(cancellationToken) ?? _connection.AcquireReadWriteSessionAsync(cancellationToken)).ConfigureAwait(false);
+                            transaction = new SpannerTransaction(_connection, TransactionMode.ReadWrite, session, null);
+
+                            TResult result = await asyncWork(transaction).ConfigureAwait(false);
+                            await transaction.CommitAsync(cancellationToken).ConfigureAwait(false);
+
+                            return result;
+                        }
+                        catch when (transaction != null)
```

Comment:
It took me a little while to work out how transaction *could* be null. Maybe add a comment?